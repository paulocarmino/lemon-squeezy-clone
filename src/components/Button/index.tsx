import * as React from 'react'
import { ImSpinner2 } from 'react-icons/im'

import clsxm from '@/utils/clsxm'

enum ButtonVariant {
  'default',
  'outline'
}

export type ButtonProps = {
  icon?: any
  block?: boolean
  isLoading?: boolean
  variant?: keyof typeof ButtonVariant
} & React.ComponentPropsWithRef<'button'>

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      children,
      className,
      disabled: buttonDisabled,
      icon = null,
      block = false,
      isLoading,
      variant = 'default',
      ...rest
    },
    ref
  ) => {
    const disabled = isLoading || buttonDisabled;

    return (
      <button
        ref={ref}
        type="button"
        disabled={disabled}
        className={clsxm(
          'relative bg-black flex items-center py-2 px-8 rounded-md hover:cursor-pointer shadow-sm',
          'focus:outline-none focus-visible:ring focus-visible:ring-primary-500 text-sm font-medium',
          'transition-colors duration-75',
          'min-h-[48px]',
          'disabled:cursor-not-allowed',
          block && 'w-full flex items-center justify-center',
          [
            variant === 'default' && ['bg-primary-500 text-white hover:bg-primary-400 disabled:bg-primary-200 disabled:hover:bg-primary-200'],
            variant === 'outline' && [
              'bg-transparent text-black border hover:border-gray-300'
            ],
          ],
          isLoading &&
          'relative text-transparent hover:text-transparent disabled:cursor-wait transition-none',
          className
        )}
        {...rest}
      >
        {icon && <div className="inline-block mr-3 text-2xl">{icon}</div>}
        {
          isLoading && (
            <div
              className={clsxm(
                'absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-black'
              )}
            >
              <ImSpinner2 className="animate-spin" data-testid="loading-icon" />
            </div>
          )
        }
        {children}
      </button >
    )
  }
)

export default Button
