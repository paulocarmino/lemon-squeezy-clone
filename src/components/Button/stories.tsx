import { Story, Meta } from '@storybook/react'
import { AiFillGoogleCircle } from 'react-icons/ai'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    block: {
      control: {
        type: 'boolean'
      }
    },
    isLoading: {
      control: {
        type: 'boolean'
      }
    },
    variant: {
      control: {
        type: 'select',
        options: ['default', 'outline']
      }
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'Sign In',
  block: false,
  isLoading: false,
  variant: 'default'
}

export const Outline: Story<ButtonProps> = (args) => <Button {...args} />

Outline.args = {
  children: 'Click to logout',
  block: false,
  isLoading: false,
  variant: 'outline'
}

export const WithIcon: Story<ButtonProps> = (args) => <Button {...args} />

WithIcon.args = {
  children: 'Sign in with Google',
  block: false,
  isLoading: false,
  variant: 'outline',
  icon: <AiFillGoogleCircle />
}
