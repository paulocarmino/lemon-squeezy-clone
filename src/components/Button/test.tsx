import { render, screen } from '@testing-library/react'
import { AiFillGoogleCircle } from 'react-icons/ai'

import Button from '.'

describe('<Button />', () => {
  it('should render default button', () => {
    const { container } = render(<Button>Sign In</Button>)

    expect(screen.getByText(/Sign In/i)).toHaveClass('py-3 px-10')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render outline button', () => {
    render(<Button variant="outline">Sign in</Button >)

    expect(screen.getByText(/Sign In/i)).toHaveClass('bg-transparent text-black border hover:border-gray-300')

  })

  it('should render icon', () => {
    render(<Button icon={<AiFillGoogleCircle data-testid="icon" />}>Sign in with Google</Button>)

    expect(screen.getByText(/Sign in with Google/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render a disabled Default Button', () => {
    render(<Button disabled>Buy now</Button>)

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveClass('disabled:bg-purple-900')
  })

  it('should render a icon when is loading', () => {
    render(<Button isLoading>Sign in with Google</Button>)

    expect(screen.getByTestId('loading-icon')).toBeInTheDocument()

  })
})
