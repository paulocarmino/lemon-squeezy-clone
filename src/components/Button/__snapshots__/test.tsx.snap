// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render default button 1`] = `
<button
  class="relative flex items-center py-2 px-8 rounded-md hover:cursor-pointer shadow-sm focus:outline-none focus-visible:ring focus-visible:ring-primary-500 text-sm font-medium transition-colors duration-75 min-h-[48px] disabled:cursor-not-allowed bg-primary-500 text-white hover:bg-primary-400 disabled:bg-primary-200 disabled:hover:bg-primary-200"
  type="button"
>
  Sign In
</button>
`;
